### PR TITLE
Implement protocol data logging

### LINKED ISSUE #107: Consider Subsystem Logger with Protocol Request and Response Data
### terraform-plugin-go version

```
v0.4.0 (not yet released)
```

### Use cases

Provider and Terraform developers may wish to see the request and response data being passed over gRPC during troubleshooting.

### Attempted solutions

Manually replacing the terraform-plugin-go dependency and injecting logging statements.

### Proposal

Create a new logging subsystem specifically for protocol request and response data at `TRACE` log level. This subsystem should default to `OFF` and should be explicitly enabled by the executor, rather infer the log level like other logging subsystems, to prevent leaking sensitive information. The details of this logging subsystem should be documented along side other logging details for this project.

### References

- https://github.com/hashicorp/terraform-plugin-go/pull/93#discussion_r709346132

### PR DESCRIPTION
**Pull request by bflad**
Closes #107

This changeset refactors the `tf5server` and `tf6server` logging implementations into a shared internal package to reduce code duplication. The protocol data handling utilizes the `TF_LOG_SDK_PROTO_DATA_DIR` environment variable to enable writing individual data field files to disk. In the future, this may need to be offered as a server configuration so it can work better in conjunction with testing frameworks (e.g. `t.TempDir()`), but there should not be too much preventing that as a later enhancement as necessary.

Without `TF_LOG_SDK_PROTO_DATA_DIR` set (once):

```text
2022-01-06T16:04:54.787-0500 [TRACE] provider.terraform-provider-framework: Skipping protocol data file writing because no data directory is set. Use the TF_LOG_SDK_PROTO_DATA_DIR environment variable to enable this functionality.: tf_req_id=105b9e19-be94-411d-2f52-ca84c14f951e tf_resource_type=framework_optional_types tf_rpc=ValidateResourceConfig @caller=/Users/bflad/src/github.com/hashicorp/terraform-plugin-go/internal/logging/protocol.go:21 @module=sdk.proto EXTRA_VALUE_AT_END=<nil> tf_provider_addr=registry.terraform.io/bflad/framework tf_proto_version=6 timestamp=2022-01-06T16:04:54.786-0500
```

With `TF_LOG_SDK_PROTO_DATA_DIR` set (per `DynamicValue` field):

```text
2022-01-06T16:06:36.079-0500 [TRACE] provider.terraform-provider-framework: Writing protocol data file: EXTRA_VALUE_AT_END=[tf_proto_data_file, 1641503196_ValidateResourceConfig_Request_Config.msgpack] tf_rpc=ValidateResourceConfig @module=sdk.proto tf_proto_version=6 tf_provider_addr=registry.terraform.io/bflad/framework tf_req_id=9084bc21-f6d1-bcc8-b916-fe75595375d2 tf_resource_type=framework_optional_types @caller=/Users/bflad/src/github.com/hashicorp/terraform-plugin-go/internal/logging/protocol.go:21 timestamp=2022-01-06T16:06:36.079-0500
```

Viewing example MessagePack file contents:

```console
$ go install github.com/nokute78/msgpack-microscope/cmd/msgpack2json@latest
$ msgpack2json 1641503196_PlanResourceChange_Request_Config.msgpack | jq .
{
  "format": "fixmap",
  "header": "0x85",
  "length": 5,
  "raw": "0x85a26964c0b36f7074696f6e616c5f74797065735f626f6f6cc0b66f7074696f6e616c5f74797065735f666c6f61743634c0b46f7074696f6e616c5f74797065735f696e743634c0b56f7074696f6e616c5f74797065735f737472696e67c0",
  "value": [
    {
      "key": {
        "format": "fixstr",
        "header": "0xa2",
        "raw": "0xa26964",
        "value": "id"
      },
      "value": {
        "format": "nil",
        "header": "0xc0",
        "raw": "0xc0",
        "value": null
      }
    },
    {
      "key": {
        "format": "fixstr",
        "header": "0xb3",
        "raw": "0xb36f7074696f6e616c5f74797065735f626f6f6c",
        "value": "optional_types_bool"
      },
      "value": {
        "format": "nil",
        "header": "0xc0",
        "raw": "0xc0",
        "value": null
      }
    },
    {
      "key": {
        "format": "fixstr",
        "header": "0xb6",
        "raw": "0xb66f7074696f6e616c5f74797065735f666c6f61743634",
        "value": "optional_types_float64"
      },
      "value": {
        "format": "nil",
        "header": "0xc0",
        "raw": "0xc0",
        "value": null
      }
    },
    {
      "key": {
        "format": "fixstr",
        "header": "0xb4",
        "raw": "0xb46f7074696f6e616c5f74797065735f696e743634",
        "value": "optional_types_int64"
      },
      "value": {
        "format": "nil",
        "header": "0xc0",
        "raw": "0xc0",
        "value": null
      }
    },
    {
      "key": {
        "format": "fixstr",
        "header": "0xb5",
        "raw": "0xb56f7074696f6e616c5f74797065735f737472696e67",
        "value": "optional_types_string"
      },
      "value": {
        "format": "nil",
        "header": "0xc0",
        "raw": "0xc0",
        "value": null
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by bflad** [comment]
Super neat, `fq` is implementing MessagePack support, which works really well with these files: https://github.com/wader/fq/issues/25

File: tfprotov5/tf5server/server.go [hunk 1]
```diff
@@ -473,98 +412,103 @@ func New(name string, serve tfprotov5.ProviderServer, opts ...ServeOpt) tfplugin
...
-		testHandle:   conf.useLoggingSink,
+		downstream:      serve,
+		stopCh:          make(chan struct{}),
+		tflogOpts:       options,
+		tflogSDKOpts:    sdkOptions,
+		name:            name,
+		useTFLogSink:    conf.useLoggingSink != nil,
+		testHandle:      conf.useLoggingSink,
+		protocolDataDir: os.Getenv(logging.EnvTfLogSdkProtoDataDir),
+		protocolVersion: "6",
```
Comment (ewbankkit) [inline tfprotov5/tf5server/server.go:429]: Is this `6` here correct given that it's in the code for a protocol v5 server?

File: tfprotov6/tf6server/server.go [hunk 2]
```diff
@@ -475,97 +414,101 @@ func New(name string, serve tfprotov6.ProviderServer, opts ...ServeOpt) tfplugin
...
-		testHandle:   conf.useLoggingSink,
+		downstream:      serve,
+		stopCh:          make(chan struct{}),
+		tflogOpts:       options,
+		tflogSDKOpts:    sdkOptions,
+		name:            name,
+		useTFLogSink:    conf.useLoggingSink != nil,
+		testHandle:      conf.useLoggingSink,
+		protocolDataDir: os.Getenv(logging.EnvTfLogSdkProtoDataDir),
+		protocolVersion: "6",
```
Comment (ewbankkit) [inline tfprotov6/tf6server/server.go:431]: Are there constants defined somewhere for protocol versions?

File: tfprotov5/tf5server/server.go (hunk 1 above)
Comment (bflad) [reply tfprotov5/tf5server/server.go:429]: Aye, it is not correct. Good eyes. Thank you.

```suggestion
		protocolVersion: "5",
```

File: tfprotov6/tf6server/server.go (hunk 2 above)
Comment (bflad) [reply tfprotov6/tf6server/server.go:431]: Not at the moment, but there certainly could be in this package. Unfortunately, the protocol buffers doesn't export that particular information. Maybe that is something that could be done there, so it is always "correct" when imported.

There are two current usages of the protocol version in these server packages:

* The string here, which is the value used for logging the protocol version.
* An int in the go-plugin instantiation, which has implications with the plugin handshake.

The challenge with a single constant is that over time the string version could and probably should include the minor protocol version, should we opt into bump that with backwards compatible new features in the protocol.

Given that, I'm tempted to introduce the following constants defined at the top of these server packages:

```go
const (
  protocolVersionMajor uint = 6
  protocolVersionMinor uint = 0
  protocolVersion = fmt.Sprintf("%d.%d", protocolVersionMajor, protocolVersionMinor)
)
```

Where `protocolVersionMajor` is used in the go-plugin handshake configuration and `protocolVersion` is saved in the server here.

**Comment by ewbankkit** [comment]
LGTM :rocket:.

```console
% go test ./...
ok  	github.com/hashicorp/terraform-plugin-go/internal/logging	0.175s
?   	github.com/hashicorp/terraform-plugin-go/tfprotov5	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/fromproto	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/toproto	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov6	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/fromproto	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/toproto	[no test files]
?   	github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server	[no test files]
ok  	github.com/hashicorp/terraform-plugin-go/tftypes	0.177s
```

**Review by alexsomesan — APPROVED** [review]
This looks great! I've also seen it in practice during our debugging session. Very cool feature.

The only thing I would suggest is maybe docs mentioning the tools used to visualise the dumped files. Fq in particular looks awesome and I was totally unaware of it.

**Comment by bflad** [comment]
> The only thing I would suggest is maybe docs mentioning the tools used to visualise the dumped files. Fq in particular looks awesome and I was totally unaware of it.

Great suggestion, I'll add some more verbiage in the README. At some point all the README documentation will get ported over to terraform.io, but no reason to skimp on content.

**Pull request merged**